src/QMCWaveFunctions/WaveFunctionComponent.cpp

```diff
@@ -22,8 +22,8 @@ namespace qmcplusplus
 // for return types
 using PsiValueType = WaveFunctionComponent::PsiValueType;
 
-WaveFunctionComponent::WaveFunctionComponent(const std::string& class_name, const std::string& obj_name)
-    : Optimizable(true),
+WaveFunctionComponent::WaveFunctionComponent(const std::string& class_name, const std::string& obj_name, bool optimizable)
+    : OptimizableObject(obj_name, optimizable),
```

Comment:
As far as I can tell we want to be able to ask a WaveFunctionComponent whether it isOptimizable() and If it is Optimizable we want to be able to call an Optimization interface corresponding to the component. 
I'd prefer something like

```
class OptimizableVars {
  virtual void checkInVariables(opt_variables_type& active) = 0;

  /** check out variational optimizable variables
   * @param active a super set of optimizable variables
   */
  virtual void checkOutVariables(const opt_variables_type& active) =0;

  /** reset the parameters during optimizations
   */
  virtual void resetParameters(const opt_variables_type& active) =0;

  /** print the state, e.g., optimizables */
  virtual void reportStatus(std::ostream& os) =0;
};

class WFC {
WFC(...) 
virtual bool isOptimizable() { return false; }
virtual OptimizableVars& getOptimizableVars() { throw LogicError; }
}

UPtrVec<OptimizableVars> extractOptimizationVars(RefVector<WFCS> wfcs) { 
  RefVector<OptimizableVars> opt_vars;
  for (auto& wfc : wfcs) {
    if wfc.isOptimizable()
      opt_vars.push_back(wfc.getOptimizationVars);
  }
  return opt_vars;
}
```

Reply:
The final product will be largely like this. The scan of optimizable parameters will be from TWF tree. That is exactly what you suggested TWF::extractOptimizationVars() and WFC::getOptimizableVars().

checkOutVariables and resetParameters are intended to be called not via TWF but directly from `RefVector<OptimizableObjects>` When this change is done, I can change `checkOutVariables/resetParameters` to pure virtual functions. The functionality of checkInVariables will be replaced with the survey function (extractOptimizationVars).

I'm going threw a few intermediate steps to make sure everything works as expected.

Reply:
I updated OptimizableObject. isOptimizable() has been moved to WFC as virtual.
All the API base implementation are std::logic_error. Their change to pure virtual needs to wait until I finish updating WFOpt drivers.